### PR TITLE
feat: Implemented local catalogs

### PR DESCRIPTION
**Pull request by quintesse**
See #180

Jbang will now look for unqualified aliases in the following order:

1. `./jbang-catalog.json` (non-hidden file!)
2. `./.jbang/jbang-catalog.json` (hidden folder!)
3. The same for any parent folder, recursively upwards
4. And finally `$HOME/.jbang/jbang-catalog.json`

Any relative paths used in those catalog files will be relative to the "base folder", so for (1) it would be `.`, but for (2) as well! The same for any parent folders. And for (4) it would be `$HOME`.

### CODE REVIEW

**Comment by maxandersen** [comment]
Why not just relative to where the catalog JSON file is ?

**Comment by zakkak** [comment]
> Why not just relative to where the catalog JSON file is ?

I think that would be more intuitive.

File: jbang-catalog.json
Comment (maxandersen) [inline jbang-catalog.json:12]: we really need to move examples out to jbang-examples but thats for a separate issue.

File: src/main/java/dev/jbang/cli/Alias.java
Comment (maxandersen) [inline src/main/java/dev/jbang/cli/Alias.java:95]: should we add a "<path catalog>" somewhere in this print out (at least when --verbose) to let users grok which catalogs are being consulted.

File: src/main/java/dev/jbang/AliasUtil.java
Comment (maxandersen) [inline src/main/java/dev/jbang/AliasUtil.java:506]: I still think we should have the actual folder of the jbang-catalog being what paths are resolved against. It just feels weird other wise.

can you point to a test/usecase that breaks/doesn't work for this ?

File: src/main/java/dev/jbang/AliasUtil.java
Comment (maxandersen) [inline src/main/java/dev/jbang/AliasUtil.java:73]: what does this do ? that smells as some very specific assumption?

**Review by maxandersen — COMMENTED** [review]
the relative path is not right to me; otherwise the feature seem to works well.

would be nice to have list print out which catalog are being consulted/included.

**Review by maxandersen — APPROVED** [review]
so lets get those changes figured out - its definitely a powerful feature!

File: src/main/java/dev/jbang/AliasUtil.java
Comment (quintesse) [reply src/main/java/dev/jbang/AliasUtil.java:73]: IT's not a specific assumption so much but more that trying to relativize  paths should come with some kind of limit, otherwise it can easily generate paths like: `../../../../../a/b/c/d/e/f.java` which is exceedingly ugly. So I just made a choice to say that 2 levels of relative paths is okay, more than that we just switch to absolute paths.

File: src/main/java/dev/jbang/cli/Alias.java
Comment (quintesse) [reply src/main/java/dev/jbang/cli/Alias.java:95]: I was actually planning on adding a `--show-origin` or `--show-location` flag to the `alias list` command (it's what `git config` does as well). But unfortunately the way I implemented this I can't give a log of how a single alias is found. (All aliases are flattened first before being looked up).

Comment (maxandersen) [reply src/main/java/dev/jbang/cli/Alias.java:95]: At least knowing the visited catalogs would be helpful. Having it per alias can be future thing if ever asked :)

File: src/main/java/dev/jbang/AliasUtil.java
Comment (zakkak) [reply src/main/java/dev/jbang/AliasUtil.java:73]: BTW, If I am not mistaken, this will not work on windows.

Comment (quintesse) [reply src/main/java/dev/jbang/AliasUtil.java:73]: It's `Path`, not `String`, it will take proper path separators into account (it's the beauty of using `Path`, you don't have to worry about stuff like that)

Comment (zakkak) [reply src/main/java/dev/jbang/AliasUtil.java:73]: Nice, thanks for the explanation @quintesse :)

**Comment by maxandersen** [comment]
bug?

```
> mkdir .jbang
> touch .jbang/jbang-catalog.json
❯ jbang alias add qs examples/progress.java
❯ jbang qs
[jbang] java.lang.IllegalArgumentException: Could not read script argument qs
[jbang] Run with --verbose for more details
```

that shouldn't fail should it ?

**Comment by maxandersen** [comment]
added commit to get better context when parse error.

**Comment by quintesse** [comment]
> that shouldn't fail should it ?

Well, the code to auto-relativize the path to a script wasn't added yet, so it expected you to explicitly put the path relative to the catalog file :-)

I just added that code, so it should work as expected now.

**Comment by quintesse** [comment]
Ok, added some tests as well. If the checks pass amd you don't find more issues this would be ok to merge for me.

**Comment by quintesse** [comment]
Don't merge yet...

**Comment by quintesse** [comment]
Ok, added a better error msg when aliases can't be properly resolved.

**Comment by maxandersen** [comment]
sorry, but with latest I'm now getting this:

```
❯ jbang alias add qs examples/progress.java
❯ jbang qs
[jbang] java.lang.IllegalArgumentException: Alias qs from /Users/max/code/personal/jbang/jbang-catalog.json failed to resolve examples/progress.java
[jbang] Run with --verbose for more details
```
and /Users/max/code/personal/jbang/examples/progress.java does exist.

**Review by maxandersen — CHANGES_REQUESTED** [review]
bug still there in different from - that needs fixing.

**Review by maxandersen — CHANGES_REQUESTED** [review]
bug still there in different from - that needs fixing.

**Comment by quintesse** [comment]
I'm not able to reproduce this :-/

**Comment by maxandersen** [comment]
debugging it and actual error is that `examples/progress.java` gets resolved to `examples/examples/progress.java`. checking why that happens.

**Comment by maxandersen** [comment]
okey, so the issue seem to be that it adds the `examples/progress.java` to the catalog but does not take into account its baseref of `examples`.

**Comment by maxandersen** [comment]
I tried modifyingtests in TestAliasNearest.java to have a test (but wont be able to complete it) that has this kind of layout:

```java
static final String baseref = "{\n" +
			"  \"aliases\": {\n" +
			"    \"baseref1\": {\n" +
			"      \"script-ref\": \"baseref1ref\"\n" +
			"    },\n" +
			"    \"dotlocal2\": {\n" +
			"      \"script-ref\": \"dotlocal2inbaseref\"\n" +
			"    },\n" +
			"    \"parent3\": {\n" +
			"      \"script-ref\": \"parent3inbaseref\"\n" +
			"    },\n" +
			"    \"global4\": {\n" +
			"      \"script-ref\": \"global4inbaseref\"\n" +
			"    }\n" +
			"  },\n" +
			"  \"base-ref\": \"test\"\n" +
			"}";
```

notice the base-ref being "test" so everything inside it is relative to "test".

Now put that file in dir outside "test" and things goes weird when you start adding to it.

**Comment by quintesse** [comment]
Aha, yes, very likely that the baseRef isn't taken into account. I'll fix that as soon as I have some time.

**Comment by quintesse** [comment]
Hmm why are the tests failing? :thinking: 
They Work On My Machine(tm)

**Comment by quintesse** [comment]
The report doesn't give much useful information either.

Could you perhaps test on your machine @maxandersen ? Just to see if it something in my environment or more the CI's environment.

**Comment by maxandersen** [comment]
Will try when back later

**Comment by maxandersen** [comment]
the tests only fail in github or using https://github.com/nektos/act
can't figure out why except I can see the scriptRef is wrong when loading catalog.

i've committed a better toString and enbabling of test output in gradle. the latter we should remove before merging though.

**Comment by quintesse** [comment]
Yay! Finally!

**Comment by maxandersen** [comment]
awesome - now I get this though (maybe not new bug ? )

```
jbang --verbose alias add -f=jbang-catalog.json test examples/lang.java
[jbang] Reading aliases from jbang-catalog.json
java.lang.NullPointerException
	at dev.jbang.AliasUtil$Aliases.getScriptBase(AliasUtil.java:126)
	at dev.jbang.AliasUtil.addAlias(AliasUtil.java:391)
	at dev.jbang.cli.AliasAdd.doCall(Alias.java:63)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:67)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:13)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1933)
	at picocli.CommandLine.access$1100(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2332)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2326)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2291)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2159)
	at picocli.CommandLine.execute(CommandLine.java:2058)
	at dev.jbang.Main.main(Main.java:14)
[jbang] If you believe this a bug in jbang open issue at https://github.com/jbangdev/jbang/issues
```

**Comment by maxandersen** [comment]
another weird one:

notice hoave gasearch@jbangdev exists but can't be resolved/found:

```
❯ jbang alias list jbangdev
env@jbangdev = Dump table of Environment Variables
gavsearch@jbangdev = Search search.maven.org for maven artifacts.
hello@jbangdev = Script that says hello back for each argument
properties@jbangdev = Dump table of System properties
❯ jbang gavsearch@jbangdev
[jbang] java.lang.IllegalArgumentException: Alias gavsearch@jbangdev from /Users/max/.jbang/cache/urls/22276b0e71b9dc0e466e9158ad103a5d9175aaadaed34eeb3bb6c8b25b80ddde/jbang-catalog.json failed to resolve gavsearch.java
[jbang] Run with --verbose for more details
❯ jbang --verbose gavsearch@jbangdev
[jbang] Downloaded catalog from https://github.com/jbangdev/jbang-catalog/blob/master/jbang-catalog.json
[jbang] Reading aliases from /Users/max/.jbang/cache/urls/22276b0e71b9dc0e466e9158ad103a5d9175aaadaed34eeb3bb6c8b25b80ddde/jbang-catalog.json
[jbang] Reading aliases from /Users/max/.jbang/jbang-catalog.json
java.lang.IllegalArgumentException: Alias gavsearch@jbangdev from /Users/max/.jbang/cache/urls/22276b0e71b9dc0e466e9158ad103a5d9175aaadaed34eeb3bb6c8b25b80ddde/jbang-catalog.json failed to resolve gavsearch.java
	at dev.jbang.cli.BaseScriptCommand.prepareScript(BaseScriptCommand.java:108)
	at dev.jbang.cli.Run.doCall(Run.java:44)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:67)
	at dev.jbang.cli.BaseCommand.call(BaseCommand.java:13)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1933)
	at picocli.CommandLine.access$1100(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2332)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2326)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2291)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2159)
	at picocli.CommandLine.execute(CommandLine.java:2058)
	at dev.jbang.Main.main(Main.java:14)
[jbang] If you believe this a bug in jbang open issue at https://github.com/jbangdev/jbang/issues
```

**Comment by quintesse** [comment]
Ok, those issues are fixed now and I added some tests for them.

**Comment by quintesse** [comment]
Again Yay! :-)

**Comment by maxandersen** [comment]
verified the two last issues I saw don't happen now - yay! going to sqaush and merge!

**Pull request closed without merging**
